Add per-event overloads to ``Sphinx.connect()``

adamtheturtle opened the pull request:
I aim to remove a `pyright` ignore comment in my source code, needed when pyright's `typeCheckingMode` is set to `"strict"`:

```python

def my_callback(app: Sphinx) -> None:
    pass

app.connect(event="builder-inited", callback=my_callback) # pyright: ignore[reportUnknownMemberType]
```

### Feature or Bugfix

- Feature

adamtheturtle committed:
Add more specific typing to application.connect callable types

I aim to remove a `pyright` ignore comment in my source code, needed when pyright's `typeCheckingMode` is set to `"strict"`:

```python

def my_callback(app: Sphinx) -> None:
    pass

app.connect(event="builder-inited", callback=my_callback) # pyright: ignore[reportUnknownMemberType]
```

danieleades reviewed (commented):
Looks pretty good to me. I've not reviewed the signatures in detail but I see a mixture of collections base classes and concrete types (eg. I see both Iterable and list used). Just checking that these are consistent and correct?

adamtheturtle committed:
Ignore a Flake8 error

adamtheturtle committed:
Avoid all but one autodoc warning - move imports out of type checking blocks

adamtheturtle committed:
Switch to 'import docutils.nodes' as this uses an existing nitpick_ignore

In conf.py there is already:

```python
nitpick_ignore = {
    ...
    ('py:class', 'docutils.nodes.TextElement'),
```

I chose to not add another ignore, and so I changed the import structure.

danieleades commented on sphinx/application.py:
```diff
@@ -13,9 +13,11 @@
 from io import StringIO
 from os import path
-from typing import IO, TYPE_CHECKING, Any, Literal
+from pathlib import Path  # NoQA: TCH003
```
Not sure about this. The preference so far has been to suppress the autodoc warning, not the lint warning.
I don't have a strong preference on which is better but we should be consistent

adamtheturtle replied:
I chose this because it is already done in this file. In particular, for example, on line 13 (just three lines up from this).

I'm open to changing it.

danieleades replied:
You're right, there are examples of both.

I think the team should take a decision on this, but it shouldn't block your PR. You're approach is locally consistent. That's good enough for now!

adamtheturtle replied:
Thank you. You can see that this change includes one `nitpick_ignore` addition.
I tried to avoid this with `import docutils.nodes` (to leverage the existing `nitpick_ignore` for `docutils.nodes.TextElement`) but there was a conflict as there is already an import of a module named `docutils`).

adamtheturtle committed:
Avoid namespace conflict and add nitpick_ignore

adamtheturtle commented on doc/conf.py:
```diff
@@ -188,6 +188,7 @@
     ('py:attr', 'srcline'),
     ('py:class', '_StrPath'),  # sphinx.environment.BuildEnvironment.doc2path
     ('py:class', 'Element'),  # sphinx.domains.Domain
+    ('py:class', 'TextElement'),
```
Some of the lines around this have a comment. Is that just the name of the first consumer? I can add `# sphinx.application.Sphinx.connect` if preferred.

AA-Turner committed:
Format & add missing

AA-Turner committed:
Fixes

AA-Turner committed:
More fixes

AA-Turner committed:
Merge branch 'master' into app.connect-specific-type

AA-Turner committed:
set -> Set

chrisjsewell reviewed (changes requested):
If you are doing this for `def connect`, then surely you also need to do it for `def emit`; otherwise you are not actually checking that they are a match?

AA-Turner reviewed (approved):
Thanks @adamtheturtle!

A

chrisjsewell reviewed (approved):
👍
